File: LeetCode_75/array_string.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class array_string {
    public void merge_strings_alternatively(String a, String b){
        int i;
        i=0;
        String result="";
        while(i<a.length() && i<b.length()){
            result+=a.charAt(i);
            result+=b.charAt(i);
            i++;
        }
        if (i<a.length()){
            result = result.concat(a.substring(i,a.length()));
        }
        if (i<b.length()){
            result = result.concat(b.substring(i,b.length()));
        }
        System.out.println(result);

    }
    public String gcdOfStrings(String str1, String str2) {  //this is actually pretty cool
        int len1 = str1.length();                           // the greatest common string divisor for 2 strings is the same
        int len2 = str2.length();                           // length as the gcd of the length of the 2 strings
        int common_den = gcd(len1, len2);
        String test  =str1.substring(0, common_den);
        String combined = str1 + str2;
        for (int i = 0; i < len1+len2; i+=common_den) {
            if (!(combined.startsWith(test, i))){
                return "";
            }
        }
        System.out.println(test);
        return test;
    }
    private int gcd(int a, int b){
        return a == 0 ? b : gcd(b % a, a);
    }

    public List<Boolean> kidsWithCandies(int[] candies, int extraCandies) {
        int max = 0;
        for (int index = 0; index < candies.length; index++) {
            if (candies[index]>max){
                max = candies[index];
            }
        }
        List<Boolean> result = new ArrayList<Boolean>();
        for (int i = 0; i < candies.length; i++) {
            if (candies[i]+extraCandies>=max){
                result.add(i, true);
            }
            else{
                result.add(i, false);
            }
        }
        return result;
    }

    public boolean canPlaceFlowers(int[] flowerbed, int n) {
        int i = 0;
        while(i<flowerbed.length){
            if (flowerbed[i]==1){
                i+=2;
            }
            else if (i!=0 && i<flowerbed.length-1){
                if (flowerbed[i-1]==0 && flowerbed[i+1]==0){
                    n--;
                    i+=2;
                }
                else if (flowerbed[i+1]!=0){
                    i+=3;
                }
                else{
                    i++;
                }
            }
            else if (i!=0 && i==flowerbed.length-1){
                if (flowerbed[i-1]==0){
                    n--;
                    i++;
                }
                else{
                    i++;
                }
            }
            else{
                if (i==flowerbed.length-1){
                    n--;
                    i++;
                }
                else if (flowerbed[i+1]==0){
                    n--;
                    i+=2;
                }
                else{
                    i+=3;
                }
            }
        }
        if (n<=0){
            return true;
        }
        else{
            return false;
        }
    }

    public String reverseVowels(String s) {// String type is immutable, so we use StringBuilder instead
        /*List<Character> vowels = new ArrayList<>(Arrays.asList('a','A','e','E','i','I','o','O','u','U'));
        ArrayList<Character> vowelsInString = new ArrayList<>();
        int i = 0;
        while (i<s.length()){
            if (vowels.contains(s.charAt(i))){
                vowelsInString.add(s.charAt(i));
            }
            i++;
        }
        i=0;
        int j = vowelsInString.size()-1;
        StringBuilder newString = new StringBuilder(s);
        while(i<s.length()){
            if (vowels.contains(s.charAt(i))){
                newString.setCharAt(i, vowelsInString.get(j));
                j--;
            }
            i++;
        }
        return newString.toString();*/
        
        //nvm i'm just stupid we can just use pointers
        List<Character> vowels = new ArrayList<>(Arrays.asList('a','A','e','E','i','I','o','O','u','U'));
        char[] stringArray = s.toCharArray();
        int start = 0;
        int end = stringArray.length-1;
        while (start<end){
            while(start<end && !(vowels.contains(stringArray[start]))){
                start++;
            }
            while(start<end && !(vowels.contains(stringArray[end]))){
                end--;
            }
            char temp = stringArray[start];
            stringArray[start]=stringArray[end];
            stringArray[end]=temp;
            start++;
            end--;
        }
        String finalResult = new String(stringArray);
        return finalResult;
    }
    
    public String reverseWords(String s) {
        String trimmedText = s.trim();
        int i = trimmedText.length()-1; //to indicate the end of the word (since we will be iterating from behind)
        int j = i; //j will iterate the string char by char until it hits a white space, which means a complete word
        StringBuilder builder = new StringBuilder();
        while (j>=0){
            if (j!=0 && trimmedText.charAt(j)!=' '){
                j--;
            }// Hello World
            else if (j==0){
                builder.append(trimmedText.substring(j, i+1));
                break;
            }
            else if (i==j){
                j--;
                i=j;
            }
            else{
                builder.append(trimmedText.substring(j+1, i+1));
                builder.append(" ");
                j--;
                i=j;
            }
        }
        String result = builder.toString();
        return result.trim();
    }

    public int[] productExceptSelf(int[] nums) {
        int[] left = new int[nums.length]; //left[i] is the product of all elements on the left of left[i]
        int[] right = new int[nums.length]; //right[i] is the product of all elements on the right of right[i]

        for (int i = 0; i < left.length; i++) {
            left[i]=1;
            right[i]=1;
        }

        for (int i = 1; i < left.length; i++) {
            left[i]=left[i-1]*nums[i-1];
        }
        
        for (int i = right.length-2; i >=0; i--) {
            right[i]=right[i+1]*nums[i+1];
        }


        int[]result = new int[nums.length];
        for (int i = 0; i < result.length; i++) {
            result[i] = left[i]*right[i];
        }
        return result;
    }

    public boolean increasingTriplet(int[] nums) {
        int min1; //tracks the smallest number so far
        long min2; //tracks the second smallest number so far
        min1=nums[0];
        min2=2147483648L;

        for (int i = 1; i < nums.length; i++) {
            if (nums[i]<min1){
                min1=nums[i];
            }
            else if(nums[i]<min2 && nums[i]>min1){
                min2=nums[i];
            }
            else if (nums[i]>min2 && min2>min1){
                return true;
            }   //while iterating, if there exists a number greater than min1 and min2, then a increasing triplet exists
        }
        return false;
    }
}

File: LeetCode_75_cpp/array_string.cpp
#include <iostream>
#include <numeric>
#include <vector>
#include <cmath>
#include <set>
using namespace std;

class Solution {
public:

//  16 Dec 2023
    string mergeAlternately(string word1, string word2) {
        int len1 = word1.length();
        int len2 = word2.length();
        string res;
        int index = 0;
        while (index < len1 && index < len2)
        {
            res += word1[index];
            res += word2[index];
            index++;
        }

        if (index < len1)
        {
            res += word1.substr(index,len1-index);
        }
        if (index < len2)
        {
            res += word2.substr(index,len2-index);
        }

        return res;
    }

//  16 Dec 2023
    string gcdOfStrings(string str1, string str2) {
        //  The length of the greatest common divisor has to the same as the gcd between the lengths of the 2 strings
        int len = gcd(str1.length(),str2.length());
        string res;
        for (int i = 0; i < len; i++)
        {
            if (str1[i] != str2[i])
            {
                return "";
            }
            res += str1[i];
        }

        string test = str1 + str2;
        for (int i = 0; i < test.length(); i += len)
        {
            if (test.substr(i,len) != res)
            {
                return "";
            }
        }
        return res;

    }

//  18 Dec 2023
    vector<bool> kidsWithCandies(vector<int>& candies, int extraCandies) {
        vector<bool> res;

        int max = 0;
        for (int &candy : candies)
        {
            if (candy > max)
            {
                max = candy;
            }
        }

        for (int &candy : candies)
        {
            if (candy+extraCandies >= max)
            {
                res.push_back(true);
            }
            else {
                res.push_back(false);
            }
        }
        return res;
    }

//  18 Dec 2023
    bool canPlaceFlowers(vector<int>& flowerbed, int n) {
        int count = 0;
        if (flowerbed.size()==1 && flowerbed[0] == 0){
            return (n<=1) ? true : false;
        }

        if (n > ceil(flowerbed.size()/2.0))
        {
            return false;
        }

        for (size_t i = 0; i < flowerbed.size(); i++)
        {
            // current location is occupied
            if (flowerbed[i]==1)
            {
                continue;
            }

            // the previous location is occupied
            if (i!=0 && flowerbed[i-1]==1)
            {
                continue;
            }

            // the next location is occupied
            if (i<flowerbed.size()-1 && flowerbed[i+1]==1)
            {
                continue;
            }

            // at location 0, and the next location is empty
            if (i==0 && flowerbed[i+1]==0)
            {
                count++;
                continue;
            }

            // at location 1, and the previous location is empty, this means we would have planted the flower at the previous location
            if (i==1 && flowerbed[0]==0)
            {
                continue;
            }

            if (i>3 && flowerbed[i-2]==0 && flowerbed[i-3]==0 && flowerbed[i-4]==1)
            {
                count++;
                continue;
            }

            if (i>4 && flowerbed[i-2]==0 && flowerbed[i-3]==0 && flowerbed[i-4]==0 && flowerbed[i-5]==0)
            {
                count++;
                continue;
            }

            if (i==3 && flowerbed[i-2]==0 && flowerbed[i-3]==0)
            {
                count++;
                continue;
            }

            if (i==2 && flowerbed[0]==0)
            {
                count++;
                continue;
            }

            // 2 plots before is occupied and the previous plot isn't, this means you can plant the flower at the current plot
            if (i>1 && flowerbed[i-2]==1)
            {
                count++;
                continue;
            }
        }
        if (count >= n)
        {
            return true;
        }
        else
        {
            return false;
        }
    }

//  22 Dec 2023
    string reverseVowels(string s) {
        if (s.length()==1)
        {
            return s;
        }
        set<char> vowels = {'a','e','i','o','u','A','E','I','O','U'};
        int left = 0;
        int right = s.length()-1;
        char temp;
        while (left<right)
        {
            if (vowels.count(s[left])==0)
            {
                left++;
                continue;
            }
            if (vowels.count(s[right])==0)
            {
                right--;
                continue;
            }
            temp = s[left];
            s[left]=s[right];
            s[right]=temp;
            left++;
            right--;
        }
        return s;
    }

//  22 Dec 2023
    string reverseWords(string s) {
        if (s.length() == 1) {
            return s;
        }

        std::string res;
        int end = static_cast<int>(s.length()) - 1;

        // Skip trailing spaces
        while (end >= 0 && s[static_cast<size_t>(end)] == ' ') {
            end--;
        }

        int start = end;

        while (start >= 0) {
            // Find the start of the word
            while (start >= 0 && s[static_cast<size_t>(start)] != ' ') {
                start--;
            }

            // Check for underflow before using start + 1
            if (start >= 0) {
                // Append the word to the result
                res += s.substr(static_cast<size_t>(start + 1), static_cast<size_t>(end - start));
                // Append a space
                res += ' ';
            }
            if (start == -1)
            {
                cout << end << endl;
                res += s.substr(0,end+2);
            }

            // Move to the next word
            end = start;

            // Skip trailing spaces
            while (end >= 0 && s[static_cast<size_t>(end)] == ' ') {
                end--;
            }

            start = end;
        }

        // Remove the trailing space before returning
        while (res[res.length()-1]==' ')
        {
            res = res.substr(0,res.length()-1);
        }
        return res;
    }

//  23 Dec 2023
    vector<int> productExceptSelf(vector<int>& nums) {
        int size = nums.size();

        //  Array of product of numbers to the left of current index
        int leftProduct[size];
        leftProduct[0] = 1;
        
        //  Array of product of numbers to the right of current index
        int rightProduct[size];
        rightProduct[size-1] = 1;
        

        for (int i = 1; i < nums.size(); i++)
        {
            leftProduct[i] = leftProduct[i-1]*nums[i-1];
        }

        for (int i = size-2; i >= 0; i--)
        {
            rightProduct[i] = rightProduct[i+1]*nums[i+1];
        }
        
        vector<int> res;
        for (int i = 0; i < size; i++)
        {
            res.push_back(leftProduct[i]*rightProduct[i]);
        }
        return res;

    }
};

int main(int argc, char *argv[]){
    Solution solution;
    vector<int> test ={1,2,3,4,5};
    vector<int> res = solution.productExceptSelf(test);
    // cout << "result: " << res << endl;
    for (int &num : res)
    {
        cout << num << "   ";
    }
    return 0;
}